internal/testtypes/matchers.go

```diff
@@ -0,0 +1,326 @@
+package testtypes
+
+import (
+	"fmt"
+
+	"github.com/onsi/gomega"
+	"github.com/onsi/gomega/gstruct"
+	"github.com/onsi/gomega/types"
+	corev1 "k8s.io/api/core/v1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Questions:
+// For composition should validFieldsWithMatchers start out as 'match all' or 'ignore all'
+
+/* Option 1:
+	Completely pre-built builders for each struct (cluster / MachineDeployment etc.
+		Completely inflexible - needs to be rebuilt for any change in the comparison
+		terse - one and done
+
+	Option 2:
+	Fully composable types - like objectbuilders
+		Field setters for each change in metadata etc.
+		Needs to be created for each test with the correct recipe
+		Many matchers may never be used
+ 		verbose
+	Option 3:
+	Fully composable types with batteries included builders / options
+		Fully composable options to add matchers etc. as in option 2
+		Additional higher level intermediate types and options added to cover most common options
+		e.g. OnlyNameNamespaceAndAPIGroupKind / excludeName for objectMeta
+*/
+
+// Comparer is a struct used to compare two objects.
+type Comparer struct {
+	// expect is the object the the Comparer will be making the comparison with. This is always a client.Object so it
+	// may sometimes be a higher order object than that covered by the Comprarer e.g. clusterv1.Cluster is passed to
+	// make comparisons for clusterv1.ClusterSpec
+	expect client.Object
+
+	// matchers is the set of matching rules that will be used in a GomegaMatcher when the Comparer is transformed.
+	matchers gstruct.Fields
+
+	// validFieldsWithMatchers is the set of validFields, and the correct GoMega for them, that applies to this
+	// specific Comparer. e.g. for TypeMeta validFieldsWithMatchers looks like:
+	// 	validFieldsWithMatchers: map[string]types.GomegaMatcher{
+	// 		"APIVersion": gomega.Equal(expect.GetObjectKind().GroupVersionKind().GroupVersion().String()),
+	// 		"Kind":       gomega.Equal(expect.GetObjectKind().GroupVersionKind().Kind),
+	//  }
+	validFieldsWithMatchers map[string]types.GomegaMatcher
+}
+
+// ClusterComparers holds the Comparers needed to assess full or custom equality between two Clusters.
+type ClusterComparers struct {
+	ObjectMeta          *Comparer
+	TypeMeta            *Comparer
+	ClusterSpecFields   *Comparer
+	ClusterStatusFields *Comparer
+}
```

Comment:
I'm wondering if we can avoid to have Type specific matcher, given that what we need is a matcher that either consider a subset of fields or ignore some fields.

Have you considered an approach similar to the one implemented in https://github.com/kubernetes-sigs/cluster-api/blob/79234da378dd31bdd2501d2be7150dfcb4851f2b/controllers/topology/internal/mergepatch/mergepatch.go#L53, that compares two generic objects (using two ways compare in this case, but we don't require it) and then filtering the diff by allow paths (restricting what we want to compare) and ignore paths (what do not want to compare in the remaining parts)?

Reply:
This is fundamental to the design for sure and it needs to be discussed. The approach I took here gives you some sort of consistency in terms of the types, but it will end up being very verbose - i.e. spec and status for each object will have to be built out. 

On the other hand using something completely generic could mean re-implementing matchers for each context / comparison they're used in. I'd like to avoid the new types but we need to give some pointers as to the comparison paths - 'validFields' here - somehow.